python/examples/tools/custom/openlibrary.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+from typing import Any
+
+import requests
+from pydantic import BaseModel, Field
+
+from beeai_framework.tools import ToolInputValidationError
+from beeai_framework.tools.tool import Tool
+
+
+class OpenLibraryToolInput(BaseModel):
+    title: str | None = Field(description="Title of book to retrieve.", default=None)
+    olid: str | None = Field(description="Open Library number of book to retrieve.", default=None)
+    subjects: str | None = Field(description="Subject of a book to retrieve.", default=None)
+
+
+class OpenLibraryToolResult(BaseModel):
+    preview_url: str
+    info_url: str
+    bib_key: str
+
+
+class OpenLibraryTool(Tool[OpenLibraryToolInput]):
+    name = "OpenLibrary"
+    description = """Provides access to a library of books with information about book titles,
+        authors, contributors, publication dates, publisher and isbn."""
+    input_schema = OpenLibraryToolInput
+
+    def _run(self, input: OpenLibraryToolInput, _: Any | None = None) -> None:
+        key = ""
+        value = ""
+        input_vars = vars(input)
+        for val in input_vars:
+            if input_vars[val] is not None:
+                key = val
+                value = input_vars[val]
+                break
+        else:
+            raise ToolInputValidationError("All input values in OpenLibraryToolInput were empty.") from None
+
+        response = requests.get(
```

Comment:
TS had emitters before and after this call, they seem unused though, is that why you left them out?

Reply:
They seemed unused and the focus seems to be more on the tool than an emitter.